src/geometricSponge.cpp

```diff
@@ -0,0 +1,333 @@
+#include "mfem/linalg/solvers.hpp"
+#include "sponge_base.hpp"
+#include "utils.hpp"
+
+using namespace mfem;
+using namespace mfem::common;
+
+GeometricSponge::GeometricSponge(mfem::ParMesh *pmesh, LoMachOptions *loMach_opts, TPS::Tps *tps)
+    : tpsP_(tps),
+      loMach_opts_(loMach_opts),
+      pmesh_(pmesh) {
+  rank_ = pmesh_->GetMyRank();
+  rank0_ = (pmesh_->GetMyRank() == 0);
+  order_ = loMach_opts->order;
+
+  tpsP_->getInput("spongeMultiplier/uniform", uniform.isEnabled, false);
+  tpsP_->getInput("spongeMultiplier/plane", plane.isEnabled, false);
+  tpsP_->getInput("spongeMultiplier/cylinder", cylinder.isEnabled, false);
+  tpsP_->getInput("spongeMultiplier/annulus", annulus.isEnabled, false);
+
+  if (uniform.isEnabled) {
+    tpsP_->getRequiredInput("spongeMultiplier/uniformMult", uniform.mult);
+  }
+
+  if (plane.isEnabled) {
+    tpsP_->getRequiredVecElem("spongeMultiplier/planeNormal", plane.normal[0], 0);
+    tpsP_->getRequiredVecElem("spongeMultiplier/planeNormal", plane.normal[1], 1);
+    tpsP_->getRequiredVecElem("spongeMultiplier/planeNormal", plane.normal[2], 2);
+    tpsP_->getRequiredVecElem("spongeMultiplier/planePoint", plane.point[0], 0);
+    tpsP_->getRequiredVecElem("spongeMultiplier/planePoint", plane.point[1], 1);
+    tpsP_->getRequiredVecElem("spongeMultiplier/planePoint", plane.point[2], 2);
+    tpsP_->getRequiredInput("spongeMultiplier/planeWidth", plane.width);
+    tpsP_->getRequiredInput("spongeMultiplier/planeMult", plane.mult);
+  }
+
+  if (cylinder.isEnabled) {
+    tpsP_->getRequiredVecElem("spongeMultiplier/cylinderPoint", cylinder.point[0], 0);
+    tpsP_->getRequiredVecElem("spongeMultiplier/cylinderPoint", cylinder.point[1], 1);
+    tpsP_->getRequiredVecElem("spongeMultiplier/cylinderPoint", cylinder.point[2], 2);
+    tpsP_->getInput("spongeMultiplier/cylinderRadiusX", cylinder.radiusX, -1.0);
+    tpsP_->getInput("spongeMultiplier/cylinderRadiusY", cylinder.radiusY, -1.0);
+    tpsP_->getInput("spongeMultiplier/cylinderRadiusZ", cylinder.radiusZ, -1.0);
+    tpsP_->getInput("spongeMultiplier/cylinderWidth", cylinder.width, 1.0e-8);
+    tpsP_->getRequiredInput("spongeMultiplier/cylinderMult", cylinder.mult);
+  }
+
+  if (annulus.isEnabled) {
+    tpsP_->getRequiredVecElem("spongeMultiplier/annulusPoint", annulus.point[0], 0);
+    tpsP_->getRequiredVecElem("spongeMultiplier/annulusPoint", annulus.point[1], 1);
+    tpsP_->getRequiredVecElem("spongeMultiplier/annulusPoint", annulus.point[2], 2);
+    tpsP_->getInput("spongeMultiplier/annulusRadiusX", annulus.radiusX, -1.0);
+    tpsP_->getInput("spongeMultiplier/annulusRadiusY", annulus.radiusY, -1.0);
+    tpsP_->getInput("spongeMultiplier/annulusRadiusZ", annulus.radiusZ, -1.0);
+    tpsP_->getInput("spongeMultiplier/annulusWidth", annulus.width, 1.0e-8);
+    tpsP_->getRequiredInput("spongeMultiplier/annulusMult", annulus.mult);
+  }
+
+  // TODO(swh): add checks and error msg for invalid sponge settings
```

Comment:
Sure.  Since the different sponge types overlap and all can be active at the same time, wasn't sure how to do the check w/o a long if with or's, which is fine, just thought you would hate the look.